src/rust/engine/graph/src/context.rs

```diff
@@ -82,14 +93,59 @@ impl<N: Node + Send> Context<N> {
     &self.inner.stats
   }
 
+  pub(crate) fn dep_record(
+    &self,
+    dep_id: EntryId,
+    generation: Generation,
+    uncacheable: bool,
+  ) -> Result<(), N::Error> {
+    let mut maybe_dep_state = self.dep_state.lock();
+    if let Some(dep_state) = maybe_dep_state.as_mut() {
+      dep_state.generations.push((dep_id, generation));
+      dep_state.has_uncacheable_deps |= uncacheable;
+      Ok(())
+    } else {
+      // This case can occur if a Node has spawned background work which continues to attempt
+      // to request dependencies in the background.
+      Err(N::Error::generic(format!(
+        "Could not request additional dependencies for {:?}: the Node has completed.",
+        self.entry_id
+      )))
+    }
+  }
+
+  ///
+  /// Gets the dependency generations which have been computed for this Node so far. May not be
+  /// called after `complete` has been called for a node.
+  ///
+  pub(crate) fn dep_generations_so_far(&self) -> Vec<(EntryId, Generation)> {
+    (*self.dep_state.lock())
+      .clone()
+      .expect("Node has already completed.")
```

Comment:
Similar to the `type_name` call recently added to `get`, is it worth putting more info into these `expect` messages for debugging (here and below)?